streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -247,8 +247,22 @@ void initializeStoreOffsetsFromCheckpoint(final boolean storeDirIsEmpty) {
                              store.changelogPartition, store.stateStore.name());
                 } else if (store.offset() == null) {
                     if (loadedCheckpoints.containsKey(store.changelogPartition)) {
-                        final Long offset = changelogOffsetFromCheckpointedOffset(loadedCheckpoints.remove(store.changelogPartition));
-                        store.setOffset(offset);
+                        final Long changelogOffset = changelogOffsetFromCheckpointedOffset(loadedCheckpoints.remove(store.changelogPartition));
+                        // StateStore#recover might roll back or forward a partially committed transaction
+                        final Long recoveredOffset;
+                        if (store.stateStore.transactional()) {
```

Comment:
Pleas see my other comment: it seems since this recover function is "best effort" anyways, we can just always restore from the checkpointed `changelogOffset`, and hence we do not need this branched logic?